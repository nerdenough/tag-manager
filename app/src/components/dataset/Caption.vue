<script setup lang="ts">
type Props = {
  caption: string
  remove: (caption: string) => void
}

defineProps<Props>()
</script>

<template>
  <div class="flex gap-x-2 bg-slate-800 text-lg p-2 rounded-md">
    <span class="flex-1">{{ caption }}</span>
    <div class="flex items-center">
      <button
        @click="() => remove(caption)"
        class="w-8 h-8 bg-slate-900 hover:bg-slate-700 text-slate-400 hover:text-red-300 rounded-md"
      >
        x
      </button>
    </div>
  </div>
</template>
