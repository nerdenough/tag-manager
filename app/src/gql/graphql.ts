/* eslint-disable */
import type { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';
export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type Dataset = {
  __typename?: 'Dataset';
  images?: Maybe<Array<Image>>;
};

export type DatasetQueries = {
  __typename?: 'DatasetQueries';
  get?: Maybe<Dataset>;
  getIdentifiers?: Maybe<Array<Scalars['String']>>;
};


export type DatasetQueriesGetArgs = {
  identifier: Scalars['String'];
};

export type Image = {
  __typename?: 'Image';
  captions?: Maybe<Array<Scalars['String']>>;
  filename: Scalars['String'];
  url: Scalars['String'];
};

export type Mutation = {
  __typename?: 'Mutation';
  updateImage: Scalars['Boolean'];
};


export type MutationUpdateImageArgs = {
  captions?: InputMaybe<Array<Scalars['String']>>;
  datasetIdentifier: Scalars['String'];
  imageFileName: Scalars['String'];
};

export type Query = {
  __typename?: 'Query';
  dataset: DatasetQueries;
};

export type GetDatasetQueryVariables = Exact<{ [key: string]: never; }>;


export type GetDatasetQuery = { __typename?: 'Query', dataset: { __typename?: 'DatasetQueries', getIdentifiers?: Array<string> | null } };

export type GetDatasetImagesQueryVariables = Exact<{
  identifier: Scalars['String'];
}>;


export type GetDatasetImagesQuery = { __typename?: 'Query', dataset: { __typename?: 'DatasetQueries', get?: { __typename?: 'Dataset', images?: Array<{ __typename?: 'Image', filename: string, url: string, captions?: Array<string> | null }> | null } | null } };


export const GetDatasetDocument = {"kind":"Document","definitions":[{"kind":"OperationDefinition","operation":"query","name":{"kind":"Name","value":"getDataset"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"dataset"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"getIdentifiers"}}]}}]}}]} as unknown as DocumentNode<GetDatasetQuery, GetDatasetQueryVariables>;
export const GetDatasetImagesDocument = {"kind":"Document","definitions":[{"kind":"OperationDefinition","operation":"query","name":{"kind":"Name","value":"getDatasetImages"},"variableDefinitions":[{"kind":"VariableDefinition","variable":{"kind":"Variable","name":{"kind":"Name","value":"identifier"}},"type":{"kind":"NonNullType","type":{"kind":"NamedType","name":{"kind":"Name","value":"String"}}}}],"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"dataset"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"get"},"arguments":[{"kind":"Argument","name":{"kind":"Name","value":"identifier"},"value":{"kind":"Variable","name":{"kind":"Name","value":"identifier"}}}],"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"images"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"filename"}},{"kind":"Field","name":{"kind":"Name","value":"url"}},{"kind":"Field","name":{"kind":"Name","value":"captions"}}]}}]}}]}}]}}]} as unknown as DocumentNode<GetDatasetImagesQuery, GetDatasetImagesQueryVariables>;