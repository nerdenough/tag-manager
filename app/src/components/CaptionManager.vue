<script setup lang="ts">
import Caption from './Caption.vue'
import { Image } from '../gql/graphql'

type Props = {
  selectedImage: Image
  modelValue: string
  addCaption: (caption: string) => void
  removeCaption: (caption: string) => void
}

const { addCaption, selectedImage } = defineProps<Props>()

const onAddCaptionFieldKeyUp = (event: KeyboardEvent) => {
  if (event.code !== 'Comma' && event.code !== 'Enter') {
    return
  }

  let caption = (event.target as HTMLInputElement).value
  if (!caption || !selectedImage.filename) {
    return
  }

  caption = event.code === 'Comma' ? caption.slice(0, -1) : caption
  addCaption(caption.trim())
}
</script>

<template>
  <div>
    <h2 class="text-xl mb-2">Captions</h2>
    <div class="flex mb-5">
      <input
        @keyup="onAddCaptionFieldKeyUp"
        :value="modelValue"
        @input="
          $emit('update:modelValue', ($event.target as HTMLInputElement).value)
        "
        type="text"
        class="flex-1 p-2 text-xl bg-slate-700 border-2 border-r-0 border-slate-700 focus:border-blue-500 hover:border-slate-600 rounded-tl-md rounded-bl-md w-64 outline-none"
        placeholder="Write a descriptive tag"
      />
      <button
        @click="() => addCaption(modelValue)"
        class="bg-blue-500 text-xl p-2 rounded-tr-md rounded-br-md"
      >
        Add
      </button>
    </div>
    <div class="flex flex-wrap gap-2">
      <Caption
        v-for="caption in selectedImage.captions"
        :caption="caption"
        :remove="removeCaption"
      />
    </div>
  </div>
</template>
