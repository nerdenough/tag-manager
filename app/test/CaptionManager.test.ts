import { mount } from '@vue/test-utils'
import Caption from '../src/components/Caption.vue'
import CaptionManager from '../src/components/CaptionManager.vue'
import { describe, it, expect } from 'vitest'
import { Image } from '../src/gql/graphql'

describe('Caption', () => {
  it('should render the component with captions for the selected image', () => {
    expect(CaptionManager).toBeTruthy()

    const selectedImage: Image = {
      filename: 'image.png',
      url: 'https://some-url/image.png',
      captions: ['some', 'captions'],
    }

    const caption = 'Some caption'
    const wrapper = mount(CaptionManager, {
      props: {
        caption,
        selectedImage,
        modelValue: 'model value',
        addCaption: () => {},
        removeCaption: () => {},
      },
    })

    expect(wrapper.find('input').element.value).toBe('model value')
    expect(wrapper.findAllComponents(Caption)).toHaveLength(2)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
