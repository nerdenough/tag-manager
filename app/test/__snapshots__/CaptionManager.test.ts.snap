// Vitest Snapshot v1

exports[`Caption > should render the component with captions for the selected image 1`] = `
"<div caption=\\"Some caption\\">
  <h2 class=\\"text-xl mb-2\\">Captions</h2>
  <div class=\\"flex mb-5\\"><input type=\\"text\\" class=\\"flex-1 p-2 font-bold text-md bg-slate-800 border-2 border-r-0 border-slate-700 focus:border-blue-500 hover:border-slate-600 rounded-tl-md rounded-bl-md w-64 outline-none\\" placeholder=\\"Write a descriptive tag\\"><button class=\\"bg-blue-500 font-bold text-md p-2 rounded-tr-md rounded-br-md\\"> Add </button></div>
  <div class=\\"flex flex-wrap gap-2\\">
    <div class=\\"flex gap-x-2 bg-slate-800 text-lg p-2 rounded-md\\"><span class=\\"flex-1\\">some</span>
      <div class=\\"flex items-center\\"><button class=\\"w-8 h-8 bg-slate-900 hover:bg-slate-700 text-slate-400 hover:text-red-300 rounded-md\\"> x </button></div>
    </div>
    <div class=\\"flex gap-x-2 bg-slate-800 text-lg p-2 rounded-md\\"><span class=\\"flex-1\\">captions</span>
      <div class=\\"flex items-center\\"><button class=\\"w-8 h-8 bg-slate-900 hover:bg-slate-700 text-slate-400 hover:text-red-300 rounded-md\\"> x </button></div>
    </div>
  </div>
</div>"
`;
